### ISSUE TITLE
bfloat16 fast_and_approximate divide: 0/0 returns 0 instead of NaN; sign-of-zero lost

### ISSUE THREAD

**Issue by dchenTT** [tenstorrent/tt-metal#43209]
### Bug

`ttnn::divide` (and the `ttnn.div` Python API by extension) produces incorrect results in two specific cases when `fast_and_approximate_mode=True` is used with `bfloat16` inputs:

1. **Division by zero**: `0.0 / 0.0` returns `0.0` instead of `NaN`. (After bf16 NaN→inf substitution that some downstream tests apply, this surfaces as `inf` expected vs. `0.0` actual.)
2. **Sign-of-zero lost**: `0.0 / -1.0` returns `+0.0` instead of `-0.0`. Small-magnitude negative quotients also lose their sign when subsequently truncated.

The bug is in the bfloat16 fast/approximate divide kernel branch — `binary_ng_device_operation.cpp` routes bf16 + `fast_and_approximate_mode=true` away from the SFPU path that handles edge cases correctly, and the alternative (likely tile-based / reciprocal-based) implementation does not preserve IEEE-754 div-by-zero / signed-zero semantics.

### Reproducer

```python
import torch
import ttnn

device = ttnn.open_device(device_id=0)
in_data1 = torch.tensor([0.0, 1.0, -1.0, 0.0, 0.0, 7.0, 9.75], dtype=torch.bfloat16)
in_data2 = torch.tensor([0.0, 0.0, 0.0, 1.0, -1.0, 2.5, -14.25], dtype=torch.bfloat16)
t1 = ttnn.from_torch(in_data1, dtype=ttnn.bfloat16, device=device, layout=ttnn.TILE_LAYOUT)
t2 = ttnn.from_torch(in_data2, dtype=ttnn.bfloat16, device=device, layout=ttnn.TILE_LAYOUT)

out = ttnn.div(t1, t2, fast_and_approximate_mode=True, rounding_mode=None)
print(ttnn.to_torch(out))
# Expected: [NaN, inf, -inf, 0., -0., 2.8, -0.6...]
# Actual:   [0.,  inf, -inf, 0.,  0.,  2.8, -0.6...]
```

The `0/0` and sign-of-zero positions diverge from torch reference.

### Where it surfaces in the test suite

`tests/ttnn/unit_tests/operations/eltwise/test_binary_fp32.py::test_binary_div_edge_case_ttnn`:

| `rounding_mode` | `fast_and_approximate_mode` | Status |
|---|---|---|
| `None` | `True` | **Skipped** since PR #29953 (2025-10-17) — see `test_binary_fp32.py:412`. The skip message documents this bug. |
| `trunc` | `True` | Currently passing only because of a workaround in `binary_composite_op.cpp` (PR #43207, this PR). |
| `floor` | `True` | Same as `trunc` — workaround. |

### Workaround currently in place

PR #43207 ("Remove use_legacy flag from eltwise binary op", closing #24162) removed legacy code that masked this bug for the `trunc`/`floor` paths. The legacy path provided two layers of safety:

1. Typecast `bfloat16 → float32` before the divide kernel (so the buggy bf16 branch was bypassed).
2. Force `fast_and_approximate_mode = std::nullopt` (so the buggy branch was never selected).

The PR removed layer 1 and partially removed layer 2. To preserve correctness, layer 2 was restored inside the trunc/floor branch of `binary_composite_op.cpp` (`div(...)` overloads, around lines 256–273 and 371–388):

```cpp
// Workaround for a known bfloat16 fast_and_approximate divide bug: 0/0 returns 0
// instead of NaN, and sign-of-zero is lost. ...
const bool suppress_fap = fast_and_approximate_mode && input.dtype() == DataType::BFLOAT16;
const bool effective_fap = suppress_fap ? false : fast_and_approximate_mode;
```

The `rounding_mode=None` path still has no workaround — only the test skip. Any caller invoking `ttnn::divide` directly (not via `ttnn::div`'s rounding-mode trunc/floor branch) still hits the bug.

### Notes for the fixer

- The fix lives at the kernel layer (in the bf16 fast/approximate divide implementation), not in `binary_composite_op.cpp`.
- Once the kernel is fixed:
  - Remove the `suppress_fap` / `effective_fap` workaround from `binary_composite_op.cpp` in BOTH `div` overloads.
  - Remove the skip in `test_binary_fp32.py:411–414` for the `bfloat16 + None + fast_and_approximate=True` case.
  - The full parametrization matrix in `test_binary_div_edge_case_ttnn` should then pass without conditional skips.
- Closed issue #26339 ("Investigate remainder and div NaN handling") covered an earlier round of div NaN cleanup but did not address this specific bf16 fast/approximate path.

### Affected ops / models

- `ttnn.div(..., fast_and_approximate_mode=True, ...)` on bf16 inputs (any rounding mode).
- `ttnn::divide` C++ API on bf16 inputs with `fast_and_approximate_mode=true`.
- Any composite op that forwards into the above.
- No production model is currently affected (Mistral 24B was the only `use_legacy=True` caller and it does not use `fast_and_approximate=True`), but the API contract is broken for any future caller that requests fap=true on bf16.

**Comment by KalaivaniMCW** [comment]
Where -0 is actually lost: **the compute datapath**

Every operation that computes a zero result drops the sign:

- 0x-1=+0: the multiply datapath does not honor the IEEE sign-of-zero rule ((+0)x(-1) should be -0).
- neg(+0) =  +0:  incorrect, because a pure sign-bit flip would give -0.

**Comment by KalaivaniMCW** [comment]
A single FPU multiply instruction has no path to emit NaN or a signed zero;  The two cases mentioned here are intrinsic to the hardware multiply unit . 
The -0 could be baked in, in the SFPU div implementation (`calculate_sfpu_binary_div`) if absolutely needed, at a perf cost for `fast_and_approximate_mode=False`. 
Also for bfloat16 dtype , `NaN`s are packed as `inf`. Any NaN in the fp32 dest is narrowed to inf by the packer. inf is the best achievable for bfloat16 dtype.

```
# Expected: [NaN, inf, -inf, 0., -0., 2.8, -0.6...]
# Actual(FPU):   [0.,  inf, -inf, 0.,  0.,  2.8, -0.6...] # fast_and_approximate_mode=True
#Actual(SFPU) : [    inf,     inf,    -inf,  0.0000,  0.0000,  2.7969, -0.6836] # fast_and_approximate_mode=False
```

**Comment by KalaivaniMCW** [comment]
Shall I close this issue @dchenTT ?
